nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py

```diff
@@ -829,193 +878,264 @@ def prompt_tuning_param_freeze_and_optimizer_setup(self):
 
         self._optimizer_param_groups = weight_decay_params, no_weight_decay_params
 
-    @classmethod
-    def _bucketize_gpt_inference(cls, batch, use_soft_prompts=False):
-        batch_tokens, lens, tokens_to_generate, compute_logprobs = batch[:4]
-        batch_size = len(batch_tokens)
-        tokens_to_generate = tokens_to_generate[0]
-        batch_tokens = batch_tokens.tolist()
-
-        if use_soft_prompts:
-            prompt_tags = batch[4]
-
-        # unpad tokens
-        indxs = [index for index in range(batch_size)]
-        for lenn, index in zip(lens, indxs):
-            batch_tokens[index] = batch_tokens[index][:lenn]
-
-        # chunk tokens by same length
-        pre_buckets, lens = [], list(set(lens.tolist()))
-        for lenn in lens:
-            pre_buckets.append([(tokens, index) for index, tokens in enumerate(batch_tokens) if len(tokens) == lenn])
-
-        buckets, positions, bucket_prompt_tags = [], [], []
-
-        # get buckets and prompts initial positions
-        for bucket in pre_buckets:
-            buckets.append(torch.tensor([item[0] for item in bucket]).to(device='cuda'))
-            positions.append([item[1] for item in bucket])
-
-            # bucket prompt tags identically to their corresponding examples
-            if use_soft_prompts:
-                bucket_prompt_tags.append([prompt_tags[item[1]] for item in bucket])
-
-        # Flatten position list
-        positions = [item for sublist in positions for item in sublist]
-
-        # Form request
-        request = {"tokens": buckets, "prompt_tags": bucket_prompt_tags}
-
-        return request, positions, tokens_to_generate, compute_logprobs[0]
-
     def get_parameters(self):
         params = []
         for param_group in self._optimizer_param_groups:
             for param in param_group['params']:
                 params.append(param)
         return params
 
-    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = None) -> Any:
-        request, positions, tokens_to_generate, compute_logprobs = MegatronGPTModel._bucketize_gpt_inference(
-            batch, self.use_soft_prompts
-        )
+    def _get_computeprob_response(self, response, inputs):
```

Comment:
Can we make this method shared as well?
The same code will be used by all Megatron NLP models, right?